### ISSUE TITLE
Code question

### ISSUE THREAD

**Issue by wuxiaofei01** [WKQ9411/Mini-LLM#2]
Thank you for the author's contribution, but there are still several issues here:

```def __getitem__(self, idx):```
The length is set to 512, which is composed of multiple pieces of different data. For example: 256 tokens of A data + <im_end> + 255 tokens of B data.This will cause confusion to the model during training. Meanwhile, why is <im_end> used here?
Network Architecture:
```
hidden_states, router_logits = decoder_layer(
                hidden_states,
                position_embeddings=position_embeddings,
                attention_mask=layer_mask,
                position_ids=position_ids,
                past_key_values=past_key_values,
                use_cache=use_cache,
                cache_position=cache_position,
                **kwargs,
            )
```
Full attention is actually used here, which is unreasonable. It will lead to the situation where each token can see both preceding and subsequent tokens during loss calculation.

**Comment by WKQ9411** [comment]
Thank you for your attention. The issues you raised:

Question 1:
> The length is set to 512, which is composed of multiple pieces of different data. For example: 256 tokens of A data + <im_end> + 255 tokens of B data.This will cause confusion to the model during training. Meanwhile, why is <im_end> used here?

It seems that what you mentioned refers to the construction of the pre-training dataset. Inserting `<|im_end|>` between different pieces of data serves two purposes: on the one hand, it separates different samples; on the other hand, during pre-training, the model needs to learn to correctly predict the next token as “end” under certain contexts. Ideally, the model’s attention should be able to automatically distinguish the samples separated by `<|im_end|>`, so that it only focuses on the context within the current sample. Of course, the most ideal approach would be to use a cross-sample mask to forcibly prevent the model from having cross-sample attention. However, in pre-training the benefit of doing so might be limited. I also consulted the nanochat maintainer Andrej Karpathy about this:
[https://github.com/karpathy/nanochat/issues/78](https://github.com/karpathy/nanochat/issues/78)

<img width="932" height="181" alt="Image" src="https://github.com/user-attachments/assets/7e03296b-69ca-4fcd-a180-9accf92d833a" />

<img width="924" height="213" alt="Image" src="https://github.com/user-attachments/assets/6122838b-b5d3-4429-a22a-25b3ffef7109" />

> However, one point needs to be clarified: when linear attention processes such concatenated data, in theory it should set the recurrent state $S$ to zero at the concatenation boundary between samples. However, its implementation is quite complex, especially during chunkwise training. I think this is actually similar to standard attention not using a cross-sample mask: the model itself should be able to learn some capability, such that it substantially rewrites the state matrix $S$ when crossing samples. In short, to simplify the implementation, we did not specifically prevent cross-sample attention during pre-training.

Question 2:
> Network Architecture: 
> ```python 
> hidden_states, router_logits = decoder_layer( ... 
> ```
> Full attention is actually used here, which is unreasonable. It will lead to the situation where each token can see both preceding and subsequent tokens during loss calculation.

The code you posted seems to be `decoder_layer` in `mini_qwen3_next`. In `decoder_layer`, different types of attention (Gated DeltaNet or Gated Attention) are applied depending on the layer. For Gated Attention, it constructs a `causal_mask` to prevent seeing future tokens. For Gated DeltaNet, it is inherently recurrent; even with chunkwise training, it also constructs a mask inside, so it should not see future tokens.

If you have any further questions, feel free to discuss!

**Comment by wuxiaofei01** [comment]
I would like to thank the author for resolving my issue, but I still have a question: why is `<im_end>` used as the delimiter instead of the original `<endoftext>`?

**Comment by WKQ9411** [comment]
I implemented the `mini_tokenizer` with reference to Qwen3's official tokenizer, so the `eos_token` is `<|im_end|>` and the `pad_token` is `<|endoftext|>`:

<div align="center">
<img width="40%" alt="Image" src="https://github.com/user-attachments/assets/caeddce3-f3eb-4e08-8b55-af193cb4b8ca">
</div>

In addition, during model initialization, there will be:

```python
self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, config.pad_token_id)
```

That is, `pad_token_id` is passed to `padding_idx`, which causes `<|endoftext|>` to have no gradient during training, refer to [nn.Embedding](https://docs.pytorch.org/docs/stable/generated/torch.nn.Embedding.html) . Therefore, if `<|endoftext|>` is used for separation, the model cannot learn any semantics related to separation. While, If `<|im_end|>` is used for separation, as an `eos_token`, it can backpropagate gradients and the model can learn its meaning of separation and the end of a sample during pre-training. In addition, during SFT, the semantics of `<|im_end|>` will be further fine-tuned to indicate the end of a message from a certain role.

The above are just my personal opinions. In fact, there are slight differences between various methods of constructing data for pre-training. I have seen some projects introduce `bos_token` during construction. Additionally, if `pad_token_id` is not passed when initializing Embedding, I think it is also possible to use `<|endoftext|>`, but during SFT, the model will need to learn the meaning of `<|im_end|>` from scratch.

**Comment by wuxiaofei01** [comment]
Thanks for your time
